build_runner/lib/src/entrypoint/run_script.dart

```diff
@@ -0,0 +1,175 @@
+import 'package:path/path.dart' as p;
+
+import '../generate/build.dart';
+import 'base_command.dart';
+import 'options.dart';
+
+class RunCommand extends BuildRunnerCommand {
+  @override
+  String get name => 'run';
+
+  @override
+  String get description =>
+      'Performs a single build, and executes a Dart script with the given arguments.';
+
+  @override
+  String get invocation =>
+      '${super.invocation.replaceFirst('[arguments]', '[build-arguments]')} '
+      '<executable> [-- [script-arguments]]';
+
+  @override
+  SharedOptions readOptions() {
+    // Here we validate that [argResults.rest] is exactly equal to all the
+    // arguments after the `--`.
+
+    var separatorPos = argResults.arguments.indexOf('--');
+
+    if (separatorPos >= 0) {
+      void throwUsageException() {
+        throw UsageException(
+            'The `run` command does not support positional args before the, '
+            '`--` separator, which should separate build args from script args.',
+            usage);
+      }
+
+      var expectedRest = argResults.arguments.skip(separatorPos + 1).toList();
+
+      // Since we expect the first argument to be the name of a script,
+      // we should skip it when comparing extra arguments.
+      var effectiveRest = argResults.rest.skip(1).toList();
+
+      if (effectiveRest.length != expectedRest.length) {
+        throwUsageException();
+      }
+
+      for (var i = 0; i < effectiveRest.length; i++) {
+        if (expectedRest[i] != effectiveRest[i]) {
+          throwUsageException();
+        }
+      }
+    }
+
+    return SharedOptions.fromParsedArgs(
+        argResults, [], packageGraph.root.name, this);
+  }
+
+  @override
+  FutureOr<int> run() async {
+    // Ensure that the user passed the name of a file to run.
+    if (argResults.rest.isEmpty) {
+      logger..severe('Must specify an executable to run.')..severe(usage);
+      return ExitCode.usage.code;
+    }
+
+    var scriptName = argResults.rest[0];
+    var passedArgs = argResults.rest.skip(1).toList();
+    var options = readOptions();
+
+    // Ensure the extension is .dart.
+    if (p.extension(scriptName) != '.dart') {
+      logger.severe(
+          '$scriptName is not a valid Dart file, and cannot be run in the VM.');
+      return ExitCode.usage.code;
+    }
+
+    // Create a temporary directory in which to execute the script.
+    var tempPath = Directory.systemTemp
+        .createTempSync('build_runner_run_script')
+        .absolute
+        .uri
+        .toFilePath();
+
+    // Create two ReceivePorts, so that we can quit when the isolate is done.
+    //
+    // Define these before starting the isolate, so that we can close
+    // them if there is a spawn exception.
+    ReceivePort onExit, onError;
+
+    // Use a completer to determine the exit code.
+    var exitCodeCompleter = Completer<int>();
+
+    try {
+      var outputMap = options.outputMap ?? {};
+      outputMap.addAll({tempPath: null});
```

Comment:
We could pass the top level directory under which the script lives here instead of null - that way only that directory would be built (it would also be hoisted up one level, which would change the path for the script you run a bit).